p2p/shim.go

```diff
@@ -0,0 +1,327 @@
+				}
+			}
+		}(cs)
+	}
+}
+
+// handlePeerErrors iterates over each p2p Channel and starts a separate go-routine
+// where we listen for peer errors. For each peer error, we find the peer from
+// the legacy p2p Switch and execute a StopPeerForError call with the corresponding
+// peer error.
+func (rs *ReactorShim) handlePeerErrors() {
+	for _, cs := range rs.Channels {
+		go func(cs *ChannelShim) {
+			for pErr := range cs.PeerErrCh {
+				if !pErr.PeerID.Empty() {
+					peer := rs.Switch.peers.Get(ID(pErr.PeerID.String()))
+					if peer == nil {
+						rs.Logger.Error("failed to handle peer error; failed to find peer", "peer", pErr.PeerID.String())
+						continue
+					}
+
+					rs.Switch.StopPeerForError(peer, pErr.Err)
+				}
+			}
+		}(cs)
+	}
+}
+
+// OnStart executes the reactor shim's OnStart hook where we start all the
+// necessary go-routines in order to proxy peer envelopes and errors per p2p
+// Channel.
+func (rs *ReactorShim) OnStart() error {
+	if rs.Switch == nil {
+		return errors.New("proxyPeerEnvelopes: reactor shim switch is nil")
+	}
+
+	// start envelope proxying and peer error handling in separate go routines
+	rs.proxyPeerEnvelopes()
+	rs.handlePeerErrors()
+
+	return nil
+}
+
+// OnStop executes the reactor shim's OnStop hook where the PeerUpdateCh is
+// closed. The caller must be sure to also stop the real reactor so the shim's
+// proxyPeerEnvelopes can exit successfully. Note, the inbound InCh is explicitly
+// closed on Receive once the real reactor stops and closes all of its p2p
+// Channels.
+func (rs *ReactorShim) OnStop() {
+	close(rs.PeerUpdateCh)
+}
+
+// GetChannel returns a p2p Channel reference for a given ChannelID. If no
+// Channel exists, nil is returned.
+func (rs *ReactorShim) GetChannel(cID ChannelID) *Channel {
+	channelShim, ok := rs.Channels[cID]
+	if ok {
+		return channelShim.Channel
+	}
+
+	return nil
+}
+
+// GetChannels implements the legacy Reactor interface for getting a slice of all
+// the supported ChannelDescriptors.
+func (rs *ReactorShim) GetChannels() []*ChannelDescriptor {
+	sortedChIDs := make([]ChannelID, 0, len(rs.Channels))
+	for cID := range rs.Channels {
+		sortedChIDs = append(sortedChIDs, cID)
+	}
+
+	sort.Slice(sortedChIDs, func(i, j int) bool { return sortedChIDs[i] < sortedChIDs[j] })
+
+	descriptors := make([]*ChannelDescriptor, len(rs.Channels))
+	for i, cID := range sortedChIDs {
+		descriptors[i] = rs.Channels[cID].Descriptor
+	}
+
+	return descriptors
+}
+
+// AddPeer sends a PeerUpdate with status PeerStatusUp on the PeerUpdateCh.
+// The embedding reactor must be sure to listen for messages on this channel to
+// handle adding a peer.
+func (rs *ReactorShim) AddPeer(peer Peer) {
+	peerID, err := PeerIDFromString(string(peer.ID()))
+	if err != nil {
+		rs.Logger.Error("failed to add peer", "peer", peer.ID(), "err", err)
+		return
+	}
+
+	rs.PeerUpdateCh <- PeerUpdate{PeerID: peerID, Status: PeerStatusUp}
```

Comment:
This will block, leading to halts and/or goroutine leaks if noone is reading from the channel. There needs to be a way for the caller to request peer updates, and to signal that it's done reading them. Similarly in `RemovePeer`.

Reply:
Technically, yes that is possible. Let's sync on that because I don't think that makes much sense imho. I want to be careful about the interactions between the shim (what will be the router to some extent) and the real reactor.

Some ideas that come to mind are using a `select` or having the shim "know" about it's real reactor counterpart and check its running state.

Reply:
Commented on this in a different thread, we can jump on a call if you'd like.

> using a select

This will drop peer updates. I feel like peer updates are probably important enough that we want to guarantee delivery of them, at the expense of requiring low-latency consumers.

> having the shim "know" about it's real reactor counterpart and check its running state

Reducing coupling is a design goal, i.e. the router shouldn't know what a reactor even is. An explicit signal mechanism (e.g. a context) is the lightest coupling I can think of.

Reply:
Yeah lets hop on a call because using a context changes a good part of this design.

Reply:
Implemented a first-class type `PeerUpdatesCh` as discussed on our call.